transport/peer/peerlist/roundrobin.go

```diff
@@ -0,0 +1,181 @@
+package peerlist
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal/errors"
+
+	"github.com/uber-go/atomic"
+)
+
+// NewRoundRobin creates a new round robin PeerList using
+func NewRoundRobin(peerIDs []transport.PeerIdentifier, agent transport.PeerAgent) (transport.PeerList, error) {
+	rr := &roundRobin{
+		peerToNode: make(map[string]*roundRobinNode, len(peerIDs)),
+		agent:      agent,
+		started:    atomic.NewBool(false),
+		nextNode:   nil,
+	}
+
+	err := rr.createPeers(peerIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	return rr, nil
+}
+
+type roundRobin struct {
+	sync.Mutex
+
+	peerToNode map[string]*roundRobinNode
+	agent      transport.PeerAgent
+	started    *atomic.Bool
+	nextNode   *roundRobinNode
+}
+
+type roundRobinNode struct {
+	peer         transport.Peer
+	nextNode     *roundRobinNode
+	previousNode *roundRobinNode
+}
+
+func (pl *roundRobin) createPeers(peerIDs []transport.PeerIdentifier) error {
+	pl.Lock()
+	defer pl.Unlock()
+	for _, peerID := range peerIDs {
+		peer, err := pl.agent.RetainPeer(peerID, pl)
+		if err != nil {
+			return err
+		}
+
+		// TODO add event/log when duplicates are inserted
+		pl.addPeer(peer)
+	}
+	return nil
+}
+
+func (pl *roundRobin) addPeer(peer transport.Peer) error {
+	_, ok := pl.peerToNode[peer.Identifier()]
+	if ok {
+		// Peer Already exists, ignore the add
+		return errors.ErrPeerAlreadyInList{
+			Peer:     peer,
+			PeerList: pl,
+		}
+	}
+
+	next := pl.nextNode
+	if next == nil {
+		// Empty List, add the first node
+		newNode := &roundRobinNode{
+			peer: peer,
+		}
+		newNode.nextNode = newNode
+		newNode.previousNode = newNode
+		pl.peerToNode[peer.Identifier()] = newNode
+
+		pl.nextNode = newNode
+		return nil
+	}
+
+	previous := next.previousNode
+
+	newNode := &roundRobinNode{
+		peer:         peer,
+		nextNode:     next,
+		previousNode: previous,
+	}
+
+	previous.nextNode = newNode
+	next.previousNode = newNode
+
+	pl.peerToNode[peer.Identifier()] = newNode
+	return nil
+}
+
+func (pl *roundRobin) removePeer(peer transport.Peer) error {
+	node, ok := pl.peerToNode[peer.Identifier()]
+	if !ok {
+		// Peer doesn't exist in the list
+		return errors.ErrPeerNotInList{
+			PeerIdentifier: peer,
+			PeerList:       pl,
+		}
+	}
+
+	if node.previousNode == node.nextNode {
+		// This is the last node, set the nextNode to this
+		pl.nextNode = nil
+		delete(pl.peerToNode, peer.Identifier())
+		return nil
+	}
+
+	if pl.nextNode == node {
+		pl.nextNode = node.nextNode
+	}
+
+	// Switch the previous node's 'next' and the next node's 'previous'
+	node.previousNode.nextNode, node.nextNode.previousNode = node.nextNode, node.previousNode
+
+	// Remove the node from our node map
+	delete(pl.peerToNode, peer.Identifier())
+
+	return nil
+}
+
+func (pl *roundRobin) nextPeer() (transport.Peer, error) {
```

Comment:
can we just drop the word Peer() from all of these function names (instead just have Add,next,clear,Choose)? I feel it's redundant peerlist.AddPeer